File: backend/app/game/ai.py
"""AI engine with minimax and alpha-beta pruning."""
import random
from typing import Optional
from .state import GameState, Player, PieceType, count_pieces, get_opponent
from .moves import Move, generate_legal_moves
from .apply_move import apply_move


# Position weights for evaluation
POSITION_WEIGHTS = [
    [0, 4, 0, 4, 0, 4, 0, 4],
    [4, 0, 5, 0, 5, 0, 5, 0],
    [0, 5, 0, 6, 0, 6, 0, 5],
    [5, 0, 6, 0, 7, 0, 6, 0],
    [0, 6, 0, 7, 0, 7, 0, 6],
    [6, 0, 7, 0, 7, 0, 6, 0],
    [0, 7, 0, 7, 0, 7, 0, 7],
    [7, 0, 7, 0, 7, 0, 7, 0],
]

# Evaluation weights
W_MATERIAL = 10
W_POSITION = 1
W_KING = 5
W_ADVANCEMENT = 0.5


def evaluate(state: GameState, ai_player: Player) -> float:
    """Evaluate board position for AI player."""
    opponent = get_opponent(ai_player)

    # Material score
    ai_pieces = count_pieces(state, ai_player)
    opp_pieces = count_pieces(state, opponent)

    material_score = (
        (ai_pieces["men"] - opp_pieces["men"]) * W_MATERIAL +
        (ai_pieces["kings"] - opp_pieces["kings"]) * (W_MATERIAL + W_KING)
    )

    # Position score
    position_score = 0
    for row in range(8):
        for col in range(8):
            piece = state.board[row][col]
            if piece:
                pos_value = POSITION_WEIGHTS[row][col]
                if piece.player == ai_player:
                    position_score += pos_value
                else:
                    position_score -= pos_value

    # Advancement score (encourage pieces to move forward)
    advancement_score = 0
    for row in range(8):
        for col in range(8):
            piece = state.board[row][col]
            if piece and piece.type == PieceType.MAN:
                if piece.player == Player.RED:
                    advancement_score += row * W_ADVANCEMENT if piece.player == ai_player else -row * W_ADVANCEMENT
                else:
                    advancement_score += (7 - row) * W_ADVANCEMENT if piece.player == ai_player else -(7 - row) * W_ADVANCEMENT

    return material_score + W_POSITION * position_score + advancement_score


def minimax(
    state: GameState,
    depth: int,
    alpha: float,
    beta: float,
    maximizing: bool,
    ai_player: Player
) -> dict:
    """Minimax with alpha-beta pruning."""
    if depth == 0 or state.game_over:
        return {"score": evaluate(state, ai_player), "move": None}

    moves = generate_legal_moves(state)

    if not moves:
        score = -10000 if maximizing else 10000
        return {"score": score, "move": None}

    best_move = moves[0]

    if maximizing:
        max_score = float("-inf")
        for move in moves:
            try:
                new_state = apply_move(state, move)
                result = minimax(new_state, depth - 1, alpha, beta, False, ai_player)

                if result["score"] > max_score:
                    max_score = result["score"]
                    best_move = move

                alpha = max(alpha, result["score"])
                if beta <= alpha:
                    break  # Beta cutoff
            except Exception:
                continue

        return {"score": max_score, "move": best_move}
    else:
        min_score = float("inf")
        for move in moves:
            try:
                new_state = apply_move(state, move)
                result = minimax(new_state, depth - 1, alpha, beta, True, ai_player)

                if result["score"] < min_score:
                    min_score = result["score"]
                    best_move = move

                beta = min(beta, result["score"])
                if beta <= alpha:
                    break  # Alpha cutoff
            except Exception:
                continue

        return {"score": min_score, "move": best_move}


def generate_explanation(move: Move, score: float, difficulty: str, state: GameState = None) -> str:
    """Generate human-readable explanation for a move."""
    reasons = []
    
    # Capture analysis
    if move.captures and len(move.captures) > 0:
        capture_count = len(move.captures)
        if capture_count > 1:
            reasons.append(f"Multi-jump! Capturing {capture_count} pieces")
        else:
            captured = move.captures[0]
            if state:
                captured_piece = state.board[captured["row"]][captured["col"]]
                if captured_piece and captured_piece.type == PieceType.KING:
                    reasons.append("Capturing opponent's king - big material gain")
                else:
                    reasons.append("Capturing opponent's piece")
            else:
                reasons.append("Capturing opponent's piece")
    
    # Promotion analysis
    if move.promotes:
        reasons.append("Promoting to king for diagonal movement in both directions")
    
    # Positional analysis
    to_row, to_col = move.to["row"], move.to["col"]
    from_row, from_col = move.from_pos["row"], move.from_pos["col"]
    
    # Check if moving towards center
    center_dist_before = abs(3.5 - from_col) + abs(3.5 - from_row)
    center_dist_after = abs(3.5 - to_col) + abs(3.5 - to_row)
    
    if center_dist_after < center_dist_before - 1:
        reasons.append("Moving towards center for more control")
    
    # Check if moving to edge (defensive)
    if to_col == 0 or to_col == 7:
        reasons.append("Moving to edge - harder to capture but less mobile")
    
    # Check advancement for regular pieces
    if state:
        piece = state.board[from_row][from_col]
        if piece and piece.type == PieceType.MAN:
            if piece.player == Player.RED:
                if to_row > from_row:
                    reasons.append("Advancing towards promotion")
            else:
                if to_row < from_row:
                    reasons.append("Advancing towards promotion")
    
    # Score-based analysis
    if score > 100:
        reasons.append("Excellent move with significant advantage")
    elif score > 50:
        reasons.append("Strong positional gain")
    elif score > 20:
        reasons.append("Good move improving position")
    
    # Safety analysis
    if state:
        # Check if the piece was in danger before
        opponent_moves = generate_legal_moves(state)
        was_in_danger = any(
            any(c["row"] == from_row and c["col"] == from_col for c in m.captures)
            for m in opponent_moves if m.captures
        )
        if was_in_danger:
            reasons.append("Escaping from capture threat")
    
    if reasons:
        return ". ".join(reasons[:3]) + "."  # Limit to 3 reasons
    elif score > 0:
        return "Solid move maintaining good position."
    else:
        return "Defensive play to stabilize position."


def get_ai_move(state: GameState, difficulty: str = "medium") -> dict:
    """Get AI's best move with difficulty setting."""
    ai_player = state.current_player
    moves = generate_legal_moves(state)

    if not moves:
        return {"move": None, "evaluation": 0, "explanation": "No moves available.", "continuation": []}

    depths = {"easy": 2, "medium": 4, "hard": 6}
    depth = depths.get(difficulty, 4)

    # Random moves for easy mode
    if difficulty == "easy" and random.random() < 0.3:
        random_move = random.choice(moves)
        return {
            "move": random_move,
            "evaluation": evaluate(state, ai_player),
            "explanation": "Exploring possibilities...",
            "continuation": []
        }

    try:
        result = minimax(state, depth, float("-inf"), float("inf"), True, ai_player)
    except Exception:
        random_move = random.choice(moves)
        return {"move": random_move, "evaluation": 0, "explanation": "Making a move...", "continuation": []}

    if not result["move"]:
        random_move = random.choice(moves)
        return {"move": random_move, "evaluation": 0, "explanation": "Making a move...", "continuation": []}

    explanation = generate_explanation(result["move"], result["score"], difficulty, state)

    # Generate continuation (next 3 moves)
    continuation = []
    temp_state = apply_move(state, result["move"])
    
    for i in range(3):
        if temp_state.game_over:
            break
        
        try:
            cont_result = minimax(temp_state, 2, float("-inf"), float("inf"), True, temp_state.current_player)
            if cont_result["move"]:
                continuation.append(cont_result["move"].model_dump())
                temp_state = apply_move(temp_state, cont_result["move"])
            else:
                break
        except Exception:
            break

    return {
        "move": result["move"],
        "evaluation": result["score"],
        "explanation": explanation,
        "continuation": continuation
    }

File: frontend/src/game/engine/ai.js
// Simple AI using Minimax with Alpha-Beta Pruning

import { cloneState, countPieces, getOpponent, PLAYERS } from './state';
import { generateLegalMoves } from './moves';
import { applyMove } from './applyMove';

/**
 * Position weights for board evaluation
 * Center control and advancement are valued
 */
const POSITION_WEIGHTS = [
  [0, 4, 0, 4, 0, 4, 0, 4],
  [4, 0, 6, 0, 6, 0, 6, 0],
  [0, 6, 0, 8, 0, 8, 0, 6],
  [4, 0, 8, 0, 10, 0, 8, 0],
  [0, 8, 0, 10, 0, 8, 0, 4],
  [6, 0, 8, 0, 8, 0, 6, 0],
  [0, 6, 0, 6, 0, 4, 0, 4],
  [4, 0, 4, 0, 4, 0, 4, 0],
];

/**
 * Evaluate board state from AI's perspective
 * Higher score = better for AI
 */
export function evaluate(state, aiPlayer) {
  const opponent = getOpponent(aiPlayer);

  // Piece counts
  const aiPieces = countPieces(state, aiPlayer);
  const oppPieces = countPieces(state, opponent);

  // Weights
  const W_PIECE = 100;
  const W_KING = 200;
  const W_POSITION = 5;
  const W_ADVANCEMENT = 3;
  const W_MOBILITY = 2;

  // Material score
  const materialScore =
    W_PIECE * (aiPieces.men - oppPieces.men) +
    W_KING * (aiPieces.kings - oppPieces.kings);

  // Position score
  let positionScore = 0;
  let advancementScore = 0;

  for (let row = 0; row < 8; row++) {
    for (let col = 0; col < 8; col++) {
      const piece = state.board[row][col];
      if (piece) {
        const posValue = POSITION_WEIGHTS[row][col];
        const advancement = aiPlayer === PLAYERS.RED ? row : 7 - row;

        if (piece.player === aiPlayer) {
          positionScore += posValue;
          if (piece.type !== 'king') {
            advancementScore += advancement;
          }
        } else {
          positionScore -= posValue;
          if (piece.type !== 'king') {
            advancementScore -= (7 - advancement);
          }
        }
      }
    }
  }

  // Mobility score (number of available moves)
  const mobilityScore = state.currentPlayer === aiPlayer
    ? generateLegalMoves(state).length
    : -generateLegalMoves(state).length;

  // Combine scores
  const totalScore =
    materialScore +
    W_POSITION * positionScore +
    W_ADVANCEMENT * advancementScore +
    W_MOBILITY * mobilityScore;

  return totalScore;
}

/**
 * Minimax with Alpha-Beta Pruning
 */
function minimax(state, depth, alpha, beta, maximizingPlayer, aiPlayer) {
  // Terminal conditions
  if (depth === 0 || state.gameOver) {
    return { score: evaluate(state, aiPlayer), move: null };
  }

  const moves = generateLegalMoves(state);

  if (moves.length === 0) {
    // No moves available - current player loses
    const score = maximizingPlayer ? -10000 : 10000;
    return { score, move: null };
  }

  let bestMove = moves[0];

  if (maximizingPlayer) {
    let maxScore = -Infinity;

    for (const move of moves) {
      try {
        const newState = applyMove(state, move);
        const result = minimax(newState, depth - 1, alpha, beta, false, aiPlayer);

        if (result.score > maxScore) {
          maxScore = result.score;
          bestMove = move;
        }

        alpha = Math.max(alpha, result.score);

        if (beta <= alpha) {
          break; // Beta cutoff
        }
      } catch (e) {
        // Skip invalid moves
        continue;
      }
    }

    return { score: maxScore, move: bestMove };
  } else {
    let minScore = Infinity;

    for (const move of moves) {
      try {
        const newState = applyMove(state, move);
        const result = minimax(newState, depth - 1, alpha, beta, true, aiPlayer);

        if (result.score < minScore) {
          minScore = result.score;
          bestMove = move;
        }

        beta = Math.min(beta, result.score);

        if (beta <= alpha) {
          break; // Alpha cutoff
        }
      } catch (e) {
        // Skip invalid moves
        continue;
      }
    }

    return { score: minScore, move: bestMove };
  }
}

/**
 * Get AI's best move
 * @param {Object} state - Current game state
 * @param {string} difficulty - 'easy', 'medium', or 'hard'
 * @returns {Object} - Best move with explanation
 */
export function getAIMove(state, difficulty = 'medium') {
  const aiPlayer = state.currentPlayer;
  const moves = generateLegalMoves(state);

  // No moves available
  if (moves.length === 0) {
    return { move: null, evaluation: 0, explanation: 'No moves available.' };
  }

  // Depth based on difficulty
  const depths = {
    easy: 2,
    medium: 4,
    hard: 6,
  };

  const depth = depths[difficulty] || 4;

  // Add some randomness for easy mode
  if (difficulty === 'easy' && Math.random() < 0.3) {
    // 30% chance to make a random move
    const randomMove = moves[Math.floor(Math.random() * moves.length)];
    return {
      move: randomMove,
      evaluation: evaluate(state, aiPlayer),
      explanation: 'Exploring different possibilities...',
    };
  }

  let result;
  try {
    result = minimax(state, depth, -Infinity, Infinity, true, aiPlayer);
  } catch (e) {
    console.error('Minimax error, falling back to random move:', e);
    const randomMove = moves[Math.floor(Math.random() * moves.length)];
    return {
      move: randomMove,
      evaluation: 0,
      explanation: 'Making a move...',
    };
  }

  // Fallback if minimax returned no move
  if (!result.move) {
    console.warn('Minimax returned no move, picking random');
    const randomMove = moves[Math.floor(Math.random() * moves.length)];
    return {
      move: randomMove,
      evaluation: 0,
      explanation: 'Making a move...',
    };
  }

  // Generate explanation
  const explanation = generateExplanation(result.move, result.score, difficulty);

  return {
    move: result.move,
    evaluation: result.score,
    explanation,
  };
}

/**
 * Generate human-readable explanation for AI's move
 */
function generateExplanation(move, score, difficulty) {
  if (!move) return 'No moves available.';

  const explanations = [];

  if (move.captures.length > 0) {
    explanations.push(`Capturing ${move.captures.length} piece${move.captures.length > 1 ? 's' : ''}.`);
  }

  if (move.promotes) {
    explanations.push('Promoting to King for more mobility.');
  }

  if (score > 200) {
    explanations.push('This move gives a strong material advantage.');
  } else if (score > 50) {
    explanations.push('Improving position and maintaining control.');
  } else if (score < -50) {
    explanations.push('Making the best of a difficult situation.');
  }

  if (explanations.length === 0) {
    explanations.push('Developing pieces toward the center.');
  }

  return explanations.join(' ');
}

/**
 * Analyze a player's move and classify it
 * @returns {Object} - Analysis with best move comparison
 */
export function analyzeMove(stateBefore, playerMove, difficulty = 'medium') {
  const player = stateBefore.currentPlayer;

  // Get AI's best move from the same position
  const aiResult = getAIMove(stateBefore, difficulty);
  const bestMove = aiResult.move;

  // Evaluate position before
  const evalBefore = evaluate(stateBefore, player);

  // Evaluate position after player's move
  const stateAfter = applyMove(stateBefore, playerMove);
  const evalAfter = evaluate(stateAfter, player);

  // Calculate evaluation delta (from player's perspective)
  const delta = evalAfter - evalBefore;

  // Classify the move
  let classification = 'good';
  let label = 'Good move';

  if (delta < -200) {
    classification = 'blunder';
    label = 'Blunder';
  } else if (delta < -100) {
    classification = 'mistake';
    label = 'Mistake';
  } else if (delta < -30) {
    classification = 'inaccuracy';
    label = 'Inaccuracy';
  }

  // Check if player made the best move
  const isBestMove = bestMove &&
    bestMove.from.row === playerMove.from.row &&
    bestMove.from.col === playerMove.from.col &&
    bestMove.to.row === playerMove.to.row &&
    bestMove.to.col === playerMove.to.col;

  if (isBestMove) {
    classification = 'best';
    label = 'Best move';
  }

  return {
    classification,
    label,
    delta,
    evalBefore,
    evalAfter,
    bestMove,
    isBestMove,
    explanation: generateAnalysisExplanation(classification, delta, bestMove, playerMove),
  };
}

/**
 * Generate explanation for move analysis
 */
function generateAnalysisExplanation(classification, delta, bestMove, playerMove) {
  if (classification === 'best') {
    return 'Excellent! This is the strongest move available.';
  }

  if (classification === 'good') {
    return 'Solid move that maintains a good position.';
  }

  if (classification === 'inaccuracy') {
    return `This move is slightly suboptimal. Consider ${formatMove(bestMove)} for a better position.`;
  }

  if (classification === 'mistake') {
    return `This loses material or position. ${formatMove(bestMove)} would have been better.`;
  }

  if (classification === 'blunder') {
    return `A significant error. ${formatMove(bestMove)} was the correct choice.`;
  }

  return '';
}

/**
 * Format a move for display
 */
function formatMove(move) {
  if (!move) return 'another move';
  const from = String.fromCharCode(97 + move.from.col) + (8 - move.from.row);
  const to = String.fromCharCode(97 + move.to.col) + (8 - move.to.row);
  return `${from}-${to}`;
}
